output_foundry.go

```diff
@@ -236,20 +239,25 @@ func (f *FoundryOutput) ValidateStateTransition(transType ChainTransitionType, n
 
 	switch transType {
 	case ChainTransitionTypeGenesis:
-		return f.checkStateGenesisTransition(semValCtx, f.MustID(), inSums, outSums)
+		return f.checkStateGenesisTransition(semValCtx, f.MustID(), outSums)
 	case ChainTransitionTypeStateChange:
 		return f.checkStateChangeTransition(next, inSums, outSums)
 	case ChainTransitionTypeDestroy:
-		if err := NativeTokenSumBalancedWithDiff(f.MustNativeTokenID(), inSums, outSums, common.Big0); err != nil {
-			return fmt.Errorf("%w: destroy foundry state transition does not balance NativeToken %s", err, f.MustNativeTokenID())
-		}
-		return nil
+		return NativeTokenInvariantsValid(f.MustNativeTokenID(), inSums, outSums, f, nil)
 	default:
 		panic("unknown chain transition type in FoundryOutput")
 	}
 }
 
-func (f *FoundryOutput) checkStateGenesisTransition(semValCtx *SemanticValidationContext, thisFoundryID FoundryID, inSums NativeTokenSum, outSums NativeTokenSum) error {
+func (f *FoundryOutput) checkStateGenesisTransition(semValCtx *SemanticValidationContext, thisFoundryID FoundryID, outSums NativeTokenSum) error {
+	// minted and melted counter must be zero at genesis
+	switch {
+	case f.MintedTokens.Cmp(common.Big0) != 0:
```

Comment:
I'm not sure this is right. According to the specs, it should be possible to mint tokens in the tx which creates the foundry, therefore MintedTokens doesn't have to be 0 at genesis. But you can't melt tokens in the same tx, because there can never be already existing tokens on the input side, so TokenDiff will never be negative for the genesis transaction of a foundry.

Reply:
Perhaps I am misreading the TIP, but given this rule https://github.com/iotaledger/tips/pull/38/files#diff-39a438c54a901ee9f1c75be884c6b90b8d2794dc799d4fd290eee3af4e28b8c6R2985 I read it as "when the foundry is new (`Current State` is empty), then `Minted` and `Melted` tokens must be 0. Did I misinterpret the TIP?3

Reply:
It might be hard to read, but it is defined above, that:
> - The **current state of the foundry** with `Foundry ID` `X` in a transaction is defined as the consumed foundry output
  where `Foundry ID` = `X`.
> - The **next state of the foundry** with `Foundry ID` `X` in a transaction is defined as the created foundry output
  where `Foundry ID` = `X`.

And that:
| A transaction that...   | Current State           | Next State               |
|-------------------------|-------------------------|--------------------------|
| Creates the foundry     | Empty                   | Output with `Foundry ID` |
| Transitions the foundry | Input with `Foundry ID` | Output with `Foundry ID` |
| Destroys the foundry    | Input with `Foundry ID` | Empty                    |


When the foundry is created, it does not appear as a consumed output in the trasanction, hence `Current State` is `empty`. `Next State` is the actual foundry output created in that transaction.

The line you referenced means that when there is no foundry output on the input side (current state empty), assume in the calculations that `Current State(Minted) = 0` and `Current State(Melted)=0`.

But it doesn't mean that `Next State(Minted)`, the actual field of the created foundry, must be `0`.

Reply:
Alright, fair enough. I was already wondering why this criteria suddenly came into play.